packages/near-membrane-base/src/membrane.ts

```diff
@@ -1401,20 +1401,48 @@ export function createMembraneMarshall() {
                     ...listOfPointersOrPrimitives: PointerOrPrimitive[]
                 ): PointerOrPrimitive => {
                     targetPointer();
-                    const fn = getSelectedTarget();
-                    const thisArg = getLocalValue(thisArgPointerOrPrimitive);
+                    // Inline getSelectedTarget().
+                    const fn = selectedTarget as Function;
+                    selectedTarget = undefined;
+                    // Inline getLocalValue().
+                    let thisArg = thisArgPointerOrPrimitive as ProxyTarget | undefined;
+                    if (typeof thisArgPointerOrPrimitive === 'function') {
+                        thisArgPointerOrPrimitive();
+                        // Inline getSelectedTarget().
+                        thisArg = selectedTarget;
+                        selectedTarget = undefined;
+                    }
                     const { length: argsLen } = listOfPointersOrPrimitives;
                     const args = new ArrayCtor(argsLen);
                     for (let i = 0, len = argsLen; i < len; i += 1) {
-                        args[i] = getLocalValue(listOfPointersOrPrimitives[i]);
+                        const pointerOrPrimitive = listOfPointersOrPrimitives[i];
+                        // Inline getLocalValue().
+                        let localValue = pointerOrPrimitive as ProxyTarget | undefined;
+                        if (typeof pointerOrPrimitive === 'function') {
+                            pointerOrPrimitive();
+                            // Inline getSelectedTarget().
+                            localValue = selectedTarget;
+                            selectedTarget = undefined;
+                        }
+                        args[i] = localValue;
                     }
                     let value;
                     try {
                         value = ReflectApply(fn, thisArg, args);
                     } catch (e: any) {
                         throw pushErrorAcrossBoundary(e);
                     }
-                    return getTransferableValue(value);
+                    // Inline getTransferableValue().
+                    if (typeof value === 'undefined') {
+                        return undefined;
+                    }
+                    if (
+                        value === null ||
+                        (typeof value !== 'function' && typeof value !== 'object')
+                    ) {
+                        return value;
+                    }
+                    return getTransferablePointer(value);
```

Comment:
So most of the cases fall into the two if statements?

Reply:
It depends. We don't know what the result of an `apply` will be. For the constructor we can likely assume an object though.